feat(babel-preset-mc-app): disable console on production

tdeekens opened the pull request:
#### Summary

This pull request adds a babel plugin to remove console.* statements on production.

#### Description

Given we are building for production we can remove console log statements. It's not strictly needed but also can not hurt much.

There are other alternatives to do this:

1. During uglification
    - Decided against it as it create a lock-in to a minifier
2. Overwriting `console.log = () => {}`
    - It's not really a good practice to overwrite built-ins
    - It also affects code which is not "ours"

Given that I went with a simple babel plugin applied only when building for prod. Moreover the babel plugin allows only overwriting specific loggers (so if needed later we can keep `error` again etc).

montezume commented:
You can use the `isRollup` variable that @emmenko introduced

tdeekens commented:
> You can use the isRollup variable that @emmenko introduced

True. Did that in the last commit.

montezume commented:
I would like to try it locally first

tdeekens commented:
> I pulled the branch, ran yarn build to build the libraries, then yarn build in playground, and started the playground in production mode, and the console logs are still there from LoggedInUserQuery, etc.

I struggle to follow the sentiment of this comment as I assume you know why we see those: they're coming from libs in `node_modules`. Are you suggesting to just use the minifier, correct?

montezume commented:
Okay, then maybe I misunderstand what we are trying to achieve with this PR. I thought the idea was to disable all console logs in production. If the idea is to only disable console logs coming from `application-products`, or whichever application is running, then yeah, this achieves that. But, it doesn't remove console logs from our other apps.

montezume commented:
Thanks for bringing this up though, I noticed that for the same reason as we are having trouble with this, we aren't removing prop-types from ours (or external) libraries!

tdeekens commented:
I tried to achieve first that log-statements from our app's are gone. So e.g. your sdk middleware. The fact that other libraries still log is a bit of a different concern to me. However, in 677b9f1 I added dropping log statements to uglify-js. This should get rid of most. However, I'd like to keep things in both places given we move to another uglifier we still retain control over our application level log statements.

tdeekens commented:
> Okay, I'm fine with merging this and then we can look into the separate babel config for vendors later.

I gave it a shot in 4b1751b. This is roughly what we would aim to achieve, right?

tdeekens commented:
@montezume sourceFolders are now excluded (even though only `node_modules` were included before).

montezume commented on packages/mc-scripts/config/create-webpack-config-for-production.js:
```diff
@@ -437,6 +437,8 @@ module.exports = ({
       {
         test: /\.js$/,
+        include: /node_modules/,
+        exclude: sourceFolders,
```
maybe we should include this in our docs. what would happen if a user passed "node_modules" as part of their sourceFolders 🤔

montezume replied:
hm, maybe we can just document it. or provide a way to override this behaviour. maybe "optimizeVendorJavascript" can be a setting?

tdeekens replied:
Sure. I can do that tomorrow.

tdeekens commented:
Alright, in the last commits 

1. I added a `enabledVendorOptimizations` flag we can pass to out webpack config
2. I simplified the concatenation a bit (to just `filter(Boolean)` cause it made head go 💥
3. I more safely defaulted toggles
   - This is a bit of a "classic" with default arguments and option objects. When having `{ a: true, b: false }` and defaulting to them. Then passing ` { b: true }` you set `a` silently to `undefined` and hence `false`. That's rarely what you want.

@emmenko and @montezume please have another look. Cheers!

emmenko commented on packages/mc-scripts/config/create-webpack-config-for-production.js:
```diff
@@ -100,91 +101,84 @@ module.exports = ({
+}) => {
+  // NOTE: Without properly defauling passing one flag (e.g. `enabledVendorOptimizations`)
+  // would disable other flags defaulted to `true` (e.g. `enableExtractCss`).
+  const defaultedToggleFlags = {
```
Nitpick: `default` and `defaulted` are very similar and it's easier to mistype them. I got myself confused for a couple of seconds. For example, in `test-utils` we name those variables `mergedXxx`. So maybe `mergedToggleFlags`?

tdeekens replied:
Sure. I don't mind. It's just how I name them generally so I am used to the naming. Will change.

emmenko replied:
Naming is hard 😇

tdeekens replied:
It's sometimes the only fun thing left in programming 😄

emmenko commented on packages/mc-scripts/config/create-webpack-config-for-production.js:
```diff
@@ -100,91 +101,84 @@ module.exports = ({
+  const defaultedToggleFlags = {
+    ...defaultToggleFlags,
+    ...toggleFlags,
+  };
```
I think we can simplify this like:

```js
const mergedToggleFlags = Object.assign({}, defaultToggleFlags, toggleFlags);
```

...and remove the default named parameter "toggleFlags ~~= defaultToggleFlags~~,"

...and we don't need the comment anymore.

tdeekens replied:
Prettier will format it to spreading syntax generally. Will see what happens if the arg is not defaulted to an object.

tdeekens commented:
@emmenko please check 35402cf

montezume commented on packages/mc-scripts/config/create-webpack-config-for-production.js:
```diff
@@ -208,248 +191,283 @@ module.exports = ({
+          include: sourceFolders,
+        },
+        /**
+         * NOTE:
```
💯

montezume commented:
Looks great now, thanks

emmenko commented:
@tdeekens would you like to merge and update the changelog PR?
